datafusion/src/logical_plan/builder.rs

```
@@ -466,12 +467,112 @@ impl LogicalPlanBuilder {
+
+                let expr = curr_plan.expressions();
+                utils::from_plan(&curr_plan, &expr, &new_inputs).unwrap()
+            }
+        }
+    }
+
     /// Apply a sort
-    pub fn sort(&self, exprs: impl IntoIterator<Item = impl Into<Expr>>) -> Result<Self> {
-        Ok(Self::from(LogicalPlan::Sort(Sort {
-            expr: normalize_cols(exprs, &self.plan)?,
-            input: Arc::new(self.plan.clone()),
-        })))
+    pub fn sort(
+        &self,
+        exprs: impl IntoIterator<Item = impl Into<Expr>> + Clone,
+    ) -> Result<Self> {
+        let schema = self.plan.schema();
+
+        // Collect sort columns that are missing in the input plan's schema
+        let mut missing_cols: Vec<Column> = vec![];
+        exprs
+            .clone()
+            .into_iter()
+            .try_for_each::<_, Result<()>>(|expr| {
+                let mut columns: HashSet<Column> = HashSet::new();
+                utils::expr_to_columns(&expr.into(), &mut columns)?;
+
+                columns.into_iter().for_each(|c| {
+                    if schema.field_from_column(&c).is_err() {
+                        missing_cols.push(c);
+                    }
+                });
+
+                Ok(())
+            })?;
+
+        if !missing_cols.is_empty() {
+            let plan = self.add_missing_columns(self.plan.clone(), &missing_cols);
+            let sort_plan = LogicalPlan::Sort(Sort {
+                expr: normalize_cols(exprs, &plan)?,
+                input: Arc::new(plan.clone()),
+            });
+            // remove pushed down sort columns
+            let mut new_expr = Vec::with_capacity(schema.fields().len());
+            schema.fields().iter().for_each(|f| {
+                let col = f.qualified_column();
+                new_expr.push(Expr::Column(col));
+            });
+            let new_schema =
+                DFSchema::new(exprlist_to_fields(&new_expr, schema).unwrap()).unwrap();
```

any reason why we want to call unwrap here instead of propagating the error back to the caller?

---

no, actually I think we should propagate the error. let me update it later.